[6.0][Macros] In-process plugin server

rintaro opened the pull request:
Cherry-pick #73725 #74734 #74785 #74941 into release/6.0

* **Explanation**: swift-syntax libraries in compiler has been complied with `-enable-library-evolution` because the toolchain provides them as a public interface for building plugins (i.e. `lib/swift/host/*.swiftmodule`). But that was not great for the performance in the compiler. This PR changes that by separating swift-syntax libs for the compiler from the one for the plugins. So the compiler and the plugin now can use different version/ABI of swift-syntax libraries from the compiler. Now swift-syntax for the compiler are compiled without library evolution.
* **Scope**: Macro plugins
* **Risk**: Mid. This changes the mechanism of in-process plugins significantly. Now in-process plugins are communicate with the compiler using the same messaging serialization with other plugins (i.e. `swift-plugin-server`, and other executable plugins), instead of directly calling the `Macro.Type` expansion method. Also, to separate the swift-syntax libraries, this introduces another set of shared swift-syntax libraries in the toolchain.
* **Testing**: Passes current test suite.
* **Issue**: rdar://130532628
* **Reviewer**: Hamish Knight (@hamishknight) Alex Hoppen (@ahoppen) Ben Barham (@bnbarham) and Daniel Rodríguez Troitiño (@drodriguez) as the author of #74785

rintaro committed:
[Macros] In-process plugin server

Separate swift-syntax libs for the compiler and for the library plugins.
Compiler communicates with library plugins using serialized messages
just like executable plugins.

* `lib/swift/host/compiler/lib_Compiler*.dylib`(`lib/CompilerSwiftSyntax`):
  swift-syntax libraries for compiler. Library evolution is disabled.
* Compiler (`ASTGen` and `swiftIDEUtilsBridging`) only depends on
  `lib/swift/host/compiler` libraries.
* `SwiftInProcPluginServer`: In-process plugin server shared library.
  This has one `swift_inproc_plugins_handle_message` entry point that
  receives a message and return the response.
* In the compiler
  * Add `-in-process-plugin-server-path` front-end option, which specifies
    the `SwiftInProcPluginServer` shared library path.
  * Remove `LoadedLibraryPlugin`, because all library plugins are managed
    by `SwiftInProcPluginServer`
  * Introduce abstract `CompilerPlugin` class that has 2 subclasses:
    * `LoadedExecutablePlugin` existing class that represents an
      executable plugin
    * `InProcessPlugins` wraps `dlopen`ed `SwiftInProcPluginServer`
  * Unified the code path in `TypeCheckMacros.cpp` and `ASTGen`, the
    difference between executable plugins and library plugins are now
    abstracted by `CompilerPlugin`

(cherry picked from commit 2f7aa428db2ab1f4b47eb4b0ffb14fb3be046b61)

rintaro committed:
[Macros] Infer in-process plugin server in swift-frontend

Not all driver can send '-in-process-plugin-server-path'. To keep
existing '-plugin-path' working, infer default server path in the
frontend.

(cherry picked from commit f08d69c6d7c30c8bd3027dd52d26af61a9377455)

rintaro committed:
[swift-syntax] Update for underscored product names

(cherry picked from commit 3f8d13d491f8ec25a09330069cb00c9cdfb4a701)

rintaro commented:
https://github.com/apple/swift-syntax/pull/2702
https://github.com/apple/swift-driver/pull/1642
https://github.com/apple/llvm-project/pull/8929
https://github.com/apple/swift-installer-scripts/pull/309
@swift-ci Please test

drodriguez commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -55,10 +55,16 @@ toolchains::GenericUnix::addPluginArguments(const ArgList &Args,
-  llvm::sys::path::append(defaultPluginPath, "host", "plugins");
+  // In-process plugin server path.
+  auto inProcPluginServerPath = pluginPath;
+  llvm::sys::path::append(inProcPluginServerPath, "host",
+                          "libSwiftInProcPluginServer.so");
```
I am afraid this is not correct for cases of cross-compilation, like when compiling a Unix target from macOS. The `.so` assumes that the target and the compiler host are the same architecture, when they are not. This needs to be `.dylib`, `.so` or `.dll` depending on the compiler host, not the target. The same is true in `DarwinToolChains.cpp` and `WindowsToolChains.cpp`.

rintaro replied:
This `SwiftInProcPluginServer` library is a "host" library for plugins, and is `dlopen()`ed  in the compiler. It must be the same platform/architecture as the compiler.

drodriguez replied:
But these files are related to the targets, not the compiler host.

I have a patch that I will try to upload tomorrow. I am not fully sure if the code around it is completely correct for every platform (for example the plugin path for Windows might want to be different, but it is part of the `WindowsToolChains.cpp`, which only applies when Windows is the target.

I think you can see the problem if from macOS you try to build something for an embedded target that uses some macro like `@DebugDescription`. The frontend will try to load the `.so` version of the library, instead of the correct `.dylib`.

rintaro replied:
Oh I see what you mean. Maybe I misunderstood the meaning of this `lib/Driver/*ToolChains.cpp`.  I will look into it deeply tomorrow. Thank you  for the heads up.

rintaro replied:
Resolved by https://github.com/swiftlang/swift/pull/74785

drodriguez committed:
[Macros] In-process plugin server library tied to compiler host, not target

PR #73725 introduced the in-process plugin server library, but the
selection of the library depends on the selected toolchain, which
depends on the compiler target, not the host. When cross-compiling (for
example from macOS to a embedded Unix target), the compiler will
incorrectly chose the `.so` file, not find it, and fail to compile
things like the `@debugDescription` macro.

Move the in-process plugin server library code from the platform
toolchains into the parent type, and code it so it uses the right name
depending on the compiler host at compilation time. This discards the
target and only relies on the compiler host for selecting the right
library.

(cherry picked from commit 55d9e7438c70e7b39e0a77a72113b66c5342b3ac)

rintaro commented:
https://github.com/apple/swift-syntax/pull/2702
https://github.com/apple/swift-driver/pull/1642
https://github.com/apple/llvm-project/pull/8929
https://github.com/apple/swift-installer-scripts/pull/309
@swift-ci Please test

rintaro committed:
[Macros] Fix memory leak

A `std::function` instance remained in memory.

rdar://131048379
(cherry picked from commit 432d5bb2fc51cabf6174e159073c1a5812f06853)

rintaro commented:
https://github.com/apple/swift-syntax/pull/2702
https://github.com/apple/swift-driver/pull/1642
https://github.com/apple/llvm-project/pull/8929
https://github.com/apple/swift-installer-scripts/pull/309
@swift-ci Please test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2702
https://github.com/apple/swift-driver/pull/1642
https://github.com/apple/llvm-project/pull/8929
https://github.com/apple/swift-installer-scripts/pull/309
preset=buildbot_incremental_linux,lsan,tools=RDA,stdlib=DA,test=no
@swift-ci Please test with preset Linux Platform

rintaro commented:
https://github.com/swiftlang/swift-syntax/pull/2702
https://github.com/swiftlang/swift-driver/pull/1642
https://github.com/apple/llvm-project/pull/8929
https://github.com/swiftlang/swift-installer-scripts/pull/309
@swift-ci Please test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2702
https://github.com/apple/swift-driver/pull/1642
https://github.com/apple/llvm-project/pull/8929
https://github.com/apple/swift-installer-scripts/pull/309
@swift-ci Please test

rintaro commented:
https://github.com/swiftlang/swift-syntax/pull/2702
https://github.com/swiftlang/swift-driver/pull/1642
https://github.com/apple/llvm-project/pull/8929
https://github.com/swiftlang/swift-installer-scripts/pull/309
@swift-ci Please test

rintaro commented:
https://github.com/swiftlang/swift-syntax/pull/2702
https://github.com/swiftlang/swift-driver/pull/1642
https://github.com/apple/llvm-project/pull/8929
https://github.com/swiftlang/swift-installer-scripts/pull/309
preset=buildbot_incremental_linux,lsan,tools=RDA,stdlib=DA,test=no
@swift-ci Please test with preset Linux Platform

rintaro commented:
Decided not to merge
